### ISSUE TITLE
[API Question] Parentheses affects declaration emit. Intentional?

### ISSUE THREAD

**Issue by AnyhowStep** [microsoft/TypeScript#32824]
Consider the following snippet,

```ts
/*
    The below types are exactly the same.
    The only difference is that one is wrapped in parentheses.
    The other is not.
*/

type InvalidKeys<K extends string|number|symbol> = { [P in K]? : never };
type InvalidKeys2<K extends string|number|symbol> = (
    { [P in K]? : never }
);

type A<T> = (
    T & InvalidKeys<"a">
);
type A2<T> = (
    T & InvalidKeys2<"a">
);

/*
    type a = {
        x: number;
    } & InvalidKeys<"a">
 */
type a = A<{ x : number }>;
/*
    type a2 = {
        x: number;
    } & {
        a?: undefined;
    }
*/
type a2 = A2<{ x : number }>;

/*
    One parentheses is all you need to change the output
*/
```

[Playground](http://www.typescriptlang.org/play/#code/PQKgUABFECoBYFMICMEBsD2B3CAXAngA4IDOEAhgE5IIAe5AxrmvnohCeQLYIB0k0eEgwA7FhAAmASwBmMhNREMkUsrjjlcEUSrJZK5QsQkQpIiISoIR60qX7RY7DLcqmyIl-xDAwYAsQQAJIiAG7kaFISANII+CQAPNEQdLjWEmQkuJRmAOYAPiIArlyolPkk+KUYaAB8EAC8EADeEADaAAqm5tEAugD8EABcECIIoQoQAL4A3P5ESCHhkTFxJABMSSm0aSIZHNl5hSVlFVXINfVNABQCUK2d3RB9gyNjE25TYACUc-OBAEEEjArhBbo4YBAAGTBMIRKKxeIJABE5GRtR+cwCSABmxBjTBd1g0NhywRa02qPRmL8oCJ2IoBOaRMctDeJwUc0cUxJS3hqyRVIxEB8-yQ5AJQNatGGow5n1qczpjgZ5HWTJZ0DZctKnKJPJhzMcjnI-RGRT2CBkZgQEi50C+otV6qauIS0tlxV1Cr+yugAHkxhYrDZECRSO4KGg0BB8BgiqMELa8BgIAwNCJckhbNoirhCHmwD4gA)

I've noticed that, sometimes, the presence/absence of parentheses can change the emitted type declaration. Is this intentional? If so, are there set rules for it? Can this behaviour be relied upon?

There are times where one would **want** the type to be expanded. And other times where one would **not** want the type to be expanded.

Having control over it would be nice. (I know I've wanted control over it plenty of times).

-----

I have encountered this discrepancy multiple times but this is the most recent example,
(that inspired me to finally make this issue).
https://github.com/microsoft/TypeScript/issues/20863#issuecomment-520551758

The `InvalidKeys` type is not expanded because I did not wrap the type in parentheses.

-----

Obligatory @keithlayne , who hates unnecessary parentheses :trolleybus:

**Comment by weswigham** [comment]
I don't think this is intentional - we normally skip past parens in anywhere where we traverse structure, I'm guessing that when we assign an alias symbol for the mapped type, we're neglecting to do so.

It's worth noting though, that even if we patch this, the following:
```ts
type InvalidKeys3<K extends string|number|symbol> = {
    "0": { [P in K]? : never }
}["0"];
```
Is probably still going to work as a way to avoid giving the mapped type an alias without changing its meaning.

**Comment by AnyhowStep** [comment]
Ooooooo

I just learned a new trick!

**Comment by weswigham** [comment]
If you put anything other than a constant in that index position, you best name that object you're indexing into, y'hear? I am _not_ responsible for how ugly your printback types get once you start immediately switch-indecing on anonymous objects.

**Comment by AnyhowStep** [comment]
@weswigham @RyanCavanaugh 

Could priority for fixing this "bug" be bumped up?
And maybe also merged into 3.5.1 after it's fixed?

I just realized that this parentheses-causing-expansion thing can actually cause really bad `.d.ts` output.

Bad enough that the project will emit `.d.ts` files fine...
But other projects using the `.d.ts` files will have a stack overflow!

```

/home/anyhowstep/node-projects/tsql/tsql/node_modules/typescript/lib/typescript.js:37580
        function getLateBoundSymbol(symbol) {
                                   ^
RangeError: Maximum call stack size exceeded
    at getLateBoundSymbol (/home/anyhowstep/node-projects/tsql/tsql/node_modules/typescript/lib/typescript.js:37580:36)
    at getSymbolOfNode (/home/anyhowstep/node-projects/tsql/tsql/node_modules/typescript/lib/typescript.js:33831:51)
    at appendTypeParameters (/home/anyhowstep/node-projects/tsql/tsql/node_modules/typescript/lib/typescript.js:36781:99)
    at getOuterTypeParameters (/home/anyhowstep/node-projects/tsql/tsql/node_modules/typescript/lib/typescript.js:36819:57)
    at getOuterTypeParameters (/home/anyhowstep/node-projects/tsql/tsql/node_modules/typescript/lib/typescript.js:36812:51)
    at getOuterTypeParameters (/home/anyhowstep/node-projects/tsql/tsql/node_modules/typescript/lib/typescript.js:36812:51)
    at getOuterTypeParameters (/home/anyhowstep/node-projects/tsql/tsql/node_modules/typescript/lib/typescript.js:36812:51)
    at getOuterTypeParameters (/home/anyhowstep/node-projects/tsql/tsql/node_modules/typescript/lib/typescript.js:36812:51)
    at getOuterTypeParameters (/home/anyhowstep/node-projects/tsql/tsql/node_modules/typescript/lib/typescript.js:36812:51)
    at getOuterTypeParameters (/home/anyhowstep/node-projects/tsql/tsql/node_modules/typescript/lib/typescript.js:36812:51)
```

**Comment by AnyhowStep** [comment]
Also, I think this parentheses thing might partially be the cause of my other emit problems like in,
https://github.com/microsoft/TypeScript/issues/31824

Because, well, I always use parentheses for complex type aliases.

-----

@keithlayne Look! A world where unnecessary parentheses crashes your code!
Your dream world! :trolleybus: 

I didn't even need 400+ parentheses this time!
